pkg/virtual/framework/forwardingregistry/store.go

```diff
@@ -291,6 +294,42 @@ func clientGetter(dynamicClusterClient dynamic.ClusterInterface, namespaceScoped
 	}
 }
 
+type listerWatcher interface {
+	List(ctx context.Context, opts metav1.ListOptions) (*unstructured.UnstructuredList, error)
+	Watch(ctx context.Context, opts metav1.ListOptions) (watch.Interface, error)
+}
+
+func listerWatcherGetter(dynamicClusterClient kcpdynamic.ClusterInterface, namespaceScoped bool, resource schema.GroupVersionResource, apiExportIdentityHash string) func(ctx context.Context) (listerWatcher, error) {
+	return func(ctx context.Context) (listerWatcher, error) {
+		cluster, err := genericapirequest.ValidClusterFrom(ctx)
+		if err != nil {
+			return nil, err
+		}
+		gvr := resource
+		clusterName := cluster.Name
+		if apiExportIdentityHash != "" {
+			gvr.Resource += ":" + apiExportIdentityHash
+		}
+		namespace, namespaceSet := genericapirequest.NamespaceFrom(ctx)
+
+		switch clusterName {
+		case logicalcluster.Wildcard:
+			if namespaceScoped && namespaceSet && namespace != metav1.NamespaceAll {
+				return nil, fmt.Errorf("cross-cluster LIST and WATCH are required to be cross-namespace, not scoped to namespace %s", namespace)
+			}
+			return dynamicClusterClient.Resource(gvr), nil
+		default:
+			if namespaceScoped && !namespaceSet {
+				if !namespaceSet {
```

Comment:
The double `!namespaceSet` looks off?

Reply:
Amazed tests did not catch that - good one!